### PR TITLE
Terms facet supports temporary base filter.

### PR DESCRIPTION
**Pull request by claygoddard**
Allow for a terms facet base filter for the purposes of determining top terms.
Sub facets/stats still use the doc set relative to the full base (not filtered base).

This allows for specialized top terms selection (will be used for funnels + retention charting)
while allowing the full doc set to be addressed in sub facets.

This is much more straightforward than trying to wrangle through domain exclusions which both require tagging
as well as re-issuing the full set of queries and filters all the way up the chain including the root
query and filters.

### CODE REVIEW

**Comment by claygoddard** [comment]
See example product usage here: https://github.com/cowpaths/fs-solr/pull/235/files#diff-6089f7b2b72e1b1bdfc94a1f1c416e368989637f02c79f2f0b26d197dd422dd6R207

It allows us to be smart about terms selection while still preserving an underlying domain that is useful for sub facets.

**Comment by claygoddard** [comment]
Interested in your thoughts (before I add some local tests). This feels like something that could be argued as a feature of terms selection, and not facet domain selection as a whole. Different facet types invoke sub facets differently and it didn't feel worth weaving through the more general facet domain configuration yet. LMK what you think.

File: solr/core/src/java/org/apache/solr/search/facet/FacetContext.java [hunk 1]
```diff
@@ -86,6 +87,14 @@ public FacetContext getParent() {
     return parent;
   }
 
+  public FacetContext nearestPermContext() {
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/search/facet/FacetContext.java]: perm? or maybe nearestParentContext?

File: solr/core/src/java/org/apache/solr/search/facet/FacetContext.java [hunk 2]
```diff
@@ -95,8 +104,26 @@ public int getFlags() {
...
+  public FacetContext subTemp(Query filter, DocSet domain) {
+    return sub(filter, domain, true);
+  }
+
+  private FacetContext sub(Query filter, DocSet domain, boolean temp) {
     FacetContext ctx = new FacetContext();
-    ctx.parent = this;
+    // We want the first non temporary context (starting with this) for parent.
+    // Typically we would only skip one context at most if temp contexts are used.
+    ctx.parent = nearestPermContext();
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/search/facet/FacetContext.java]: just to add, looks like nearestPermContext can return null as well ?

**Review by hiteshk25 — APPROVED** [review]
LGTM

File: solr/core/src/java/org/apache/solr/search/facet/FacetContext.java (hunk 1 above)
Comment (claygoddard) [reply solr/core/src/java/org/apache/solr/search/facet/FacetContext.java]: Yeah, I can just simplify and call it `perm`. That SGTM.

File: solr/core/src/java/org/apache/solr/search/facet/FacetContext.java (hunk 2 above)
Comment (claygoddard) [reply solr/core/src/java/org/apache/solr/search/facet/FacetContext.java]: It theoretically could if the root was temporary and we were becoming a child off of it. But that's what we would want - works fine everywhere.

**Pull request merged**